x/fileurl.go

```diff
@@ -0,0 +1,110 @@
+package x
+
+import (
+	"net/url"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/ory/x/logrusx"
+)
+
+// winPathRegex is a regex for [DRIVE-LETTER]:
+var winPathRegex = regexp.MustCompile("^[A-Za-z]:.*")
+
+// GetURLFilePath returns the path of a URL that is compatible with the runtime os filesystem
+func GetURLFilePath(u *url.URL) string {
+	if u == nil {
+		return ""
+	}
+
+	if u.Scheme != "file" && u.Scheme != "" {
+		return u.Path
+	}
+	fPath := u.Path
+	if runtime.GOOS == "windows" {
+		if u.Host != "" {
+			// Make UNC Path
+			s := string(filepath.Separator)
+			fPath = s + s + u.Host + filepath.FromSlash(fPath)
+			return fPath
+		}
+		if winPathRegex.MatchString(fPath[1:]) {
+			// On Windows we should remove the initial path separator in case this
+			// is a normal path (for example: "\c:\" -> "c:\"")
+			fPath = stripFistPathSeparators(fPath)
+		}
+	}
+	return filepath.FromSlash(fPath)
+}
+
+// ParseURL parses rawurl into a URL structure with special handling for file:// URLs
+func ParseURL(rawurl string) (*url.URL, error) {
+	lcRawurl := strings.ToLower(rawurl)
```

Comment:
Wouldn't it be less error prone to do
```suggestion
        if runtime.GOOS == "windows" {
	    rawURL = strings.ToLower(rawurl)
	}
```
and avoid the decision what variable to use for what check?

Reply:
runtime.GOOS is guaranteed to always be lower case, so I don't think this is needed. 
If you run `go tool dist list` you will see a list of all combinations of runtime.GOOS and runtime.GOARCH.

Reply:
you are lower casing `rawURL` there because windows filepaths are case insensitive right? But they are case sensitive on POSIX
That was what I was referring to.

Reply:
Not sure exactly what you mean here, the lcRawURL is only used to compare the initial part if the rawURL string so see if it starts with file://  (or any upper/lower case version of it).

Reply:
Ah ok, could you then maybe use `strings.Split(rawURL, "://")` to get the scheme and use that for the checks?

Reply:
At this point in the code I am first looking for file:/// (three slashes), if that is the case I am parsing it as a normal URL with an early exit. If that is not the case and it instead starts with file:// (two slashes) we trim the first 7 chars off the beginning and then continues the function. The lower case comparison has nothing to do with Windows, it just makes sure that FILE:// and file:// are treated the same.
I can change it to a regexp case insensitive comparison if that feels better.